### ISSUE TITLE
MPS `any()` crashes on a Tensor with >4 dims

### ISSUE THREAD

**Issue by Jonas-Metzger** [pytorch/pytorch#136134]
### 🐛 Describe the bug

On apple silicon, this command:

```
python -c "import torch; print(torch.randn(2,1,1,1,1).to('mps').isnan().any())"
```
throws:

```
Assertion failed: (0 <= mpsAxis && mpsAxis < 4 && "Runtime canonicalization must simplify reduction axes to minor 4 dimensions."), function getKernelAxes, file GPUReductionOps.mm, line 31.
zsh: abort      python -c 
```

It seems that `.any()` would need to implement an automatic reshape under the hood for mps?

### Versions

Versions

PyTorch version: 2.4.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.4 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.6 | packaged by conda-forge | (main, Sep 11 2024, 04:55:15) [Clang 17.0.6 ] (64-bit runtime)
Python platform: macOS-14.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Pro

Versions of relevant libraries:
[pip3] numpy==2.1.1
[pip3] torch==2.4.1
[pip3] torchaudio==2.4.1
[pip3] torchvision==0.19.1
[conda] numpy 2.1.1 py312h801f5e3_0 conda-forge
[conda] pytorch 2.4.1 py3.12_0 pytorch
[conda] torchaudio 2.4.1 py312_cpu pytorch
[conda] torchvision 0.19.1 py312_cpu pytorch

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

**Comment by janeyx99** [comment]
Can repro on 2.4.1, crashes for any tensor with more than 4 dimensions.

**Comment by malfet** [comment]
I have a strange sense of deja-vu about it (i.e. I've fixed one of those in the past, see https://github.com/pytorch/pytorch/issues/114942 )

**Comment by Jonas-Metzger** [comment]
Thanks for taking this on! FYI, I just updated to MacOS 15 RC, and the issue persists. 

Note that it doesn't occur if all tensor dims are equal to 1. At least one dim of size>1 is required.

**Comment by malfet** [comment]
@Jonas-Metzger interesting. I've noticed that it's not present in trunk builds, and assumed it was due to OS update, let me try to identify the commit that fixed it

**Comment by malfet** [comment]
And indeed it sounded familiar to me, as I've fixed in in the past by https://github.com/pytorch/pytorch/pull/130542

**Comment by atalman** [comment]
Validated, fixed in 2.5.1 final rc:
```
python -c "import torch; print(torch.__version__, torch.randn(2,1,1,1,1).to('mps').isnan().any())"
2.5.0 tensor(False, device='mps:0')
```
